src/cpp/session/SessionRpc.cpp

```diff
@@ -332,6 +381,24 @@ Error initialize()
 
    RS_REGISTER_CALL_METHOD(rs_invokeRpc);
 
+   // TODO: asyncRpc - add option to registerRpcMethod?
+   s_offlineableUris.insert("/rpc/save_document");
```

Comment:
Is there a way we can further validate that we aren't unintentionally introducing any issues with having these specific RPCs run on separate threads? I believe we'll have caught the "obvious" ones (interacting with R from a separate thread will cause immediate crashes) but worry about more subtle issues that might sneak by.

I wonder if a local development build with thread sanitizers enabled could help catch any other potential issues: https://clang.llvm.org/docs/ThreadSanitizer.html

Are we considering this, or any part of these changes, for 1.4-JR? If we are, I wonder if we could take the async RPC work but leave the offline-service work for 1.5, as it seems a bit more risky. Does that make sense?

Reply:
I will give it a go but suspect it might have too many false positives. It's a pretty hard problem to identify meaningful race conditions. We could have one big mutex around the logic in "handleConnection" but that might cause more problems than it's worth?  For now, I've been carefully reviewing the code in each offlineable request but you are right that it's very possible to miss something. There could be race conditions or also buried references to r::exec features.

Right now, this is scheduled only for 1.5 but I agree that the async-rpc stuff is a lot clearer cut and makes a big difference just by itself.

Reply:
Unfortunately, the thread sanitizer does have lots of false positives. I only looked into one in rserver where it complained about a thread race involving boost regex's use of a buffer pool from two different threads, even though from reading the code,  only one thread would be using one buffer at a time. In the docs, they say it doesn't work for C++11 mode very well so that might be a contributing factor.

Reply:
That is unfortunate! Thanks for taking the time to give that a whirl -- if all we saw was false positives, I think we can still take that as a good sign (assuming that false negatives would be rare).